### PR TITLE
Correctly propagate failures while update the flow-controller to the …

### PR DESCRIPTION
**Pull request by normanmaurer**
…multiplexed channel

Motivation:

We may fail to update the flow-controller and in this case need to notify the stream channel and close it.

Modifications:

Attach a future to the write of the update frame and in case of a failure propagate it to the channel and close it

Result:

Fixes https://github.com/netty/netty/issues/9663

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ti2ger92 PTAL

**Comment by normanmaurer** [comment]
This still needs a test case....

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 1]
```diff
@@ -817,7 +834,11 @@ private void updateLocalWindowIfNeeded() {
             if (flowControlledBytes != 0) {
                 int bytes = flowControlledBytes;
                 flowControlledBytes = 0;
-                write0(parentContext(), new DefaultHttp2WindowUpdateFrame(bytes).stream(stream));
+                write0(parentContext(), new DefaultHttp2WindowUpdateFrame(bytes).stream(stream))
+                        // Add a listener which will notify and teardown the stream
+                        // when a window update fails.
+                        // See https://github.com/netty/netty/issues/9663
+                        .addListener(WINDOW_UPDATE_FRAME_LISTENER);
```
Comment (carl-mastrangelo) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: Should the ChannelFuture be attached to the child channel?

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: @carl-mastrangelo not sure I follow. The returned `ChannelFuture` belongs to the child channel.

Comment (carl-mastrangelo) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: What I mean is shouldn't this just teardown the stream channel, not the parent channel?   If there was a failure applying the window for the child, it seems only the child should be torn down.

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: @carl-mastrangelo yes this was what it supposed to do but I messed up 🤦‍♂ ... Good catch.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 2]
```diff
@@ -105,6 +105,28 @@ public Handle newHandle() {
...
+        @Override
+        public void operationComplete(ChannelFuture future) {
+            handleWindowUpdateFrameFuture(future);
+        }
+    };
+
+    private void handleWindowUpdateFrameFuture(ChannelFuture future) {
+        Throwable cause = future.cause();
+        if (cause != null) {
+            cause.printStackTrace();
```
Comment (rhernandez35) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: Maybe log this instead of dumping it to stderr

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: actually this is a left-over from debugging. This should are just removed.

**Comment by bryce-anderson** [comment]
This seems fine to me but I'm having a hard time figuring out why it solves the core issue in #9663 which is that it appears we've 'processed' more bytes than we received.

**Comment by normanmaurer** [comment]
@bryce-anderson I am still trying to understand that as well 🤦‍♂

**Comment by normanmaurer** [comment]
@bryce-anderson alright ... I found out why it fixed the problem and re-factored to still batch the window update writes. The problem was some re-entrancy problem. See the comments in the code.

**Comment by normanmaurer** [comment]
@ti2ger92 please provide a standalone reproducer (netty only).  Keep in mind that window-updates will only be send when really needed (the window was used almost completely ).

**Comment by ti2ger92** [comment]
@normanmaurer , when the change you made is applied to our system, our use case seems OK. I think you fixed it 🎉🎊🎉🏁🎉

Can you clarify how the gist I uploaded should be reduced further?

**Comment by normanmaurer** [comment]
@ti2ger92 a testcase / reproducer that only uses netty would be nice :)

**Comment by normanmaurer** [comment]
@ti2ger92 thanks for testing it

**Pull request merged**